### PR TITLE
Johluo/targeting pack tests

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/dotnet/aspnetcore/issues/19151 and https://github.com/dotnet/aspnetcore/issues/14832. Note that I'm making the changes to 5.0 first. I'll port to 3.1 when it's done.

### CODE REVIEW

File: src/Framework/test/TargetingPackTests.cs [hunk 1]
```diff
@@ -26,12 +26,37 @@ public TargetingPackTests(ITestOutputHelper output)
...
+            Assert.All(dlls, path =>
+            {
+                var assemblyName = AssemblyName.GetAssemblyName(path);
+                using var fileStream = File.OpenRead(path);
+                using var peReader = new PEReader(fileStream, PEStreamOptions.Default);
+                var reader = peReader.GetMetadataReader(MetadataReaderOptions.Default);
+                var assemblyDefinition = reader.GetAssemblyDefinition();
+
+                Assert.True(
+                    assemblyDefinition.Version.Revision == 0 && assemblyDefinition.Version.Revision == 0,
```
Comment (wtgodbe) [inline src/Framework/test/TargetingPackTests.cs:55]: Why are these both `Revision`?

**Review by wtgodbe — APPROVED** [review]
Looks good other than the 1 question

File: src/Framework/test/TargetingPackTests.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/Framework/test/TargetingPackTests.cs:55]: It's a typo, should be build number version.

File: src/Framework/test/TargetingPackTests.cs [hunk 2]
```diff
@@ -26,12 +26,37 @@ public TargetingPackTests(ITestOutputHelper output)
...
+                    $"{path} should have a 0.0 revision number and build number version, e.g. major.minor.0.0");
```
Comment (dougbu) [inline src/Framework/test/TargetingPackTests.cs:56]: Suggest adding more detail. The message won't tell us what assembly is busted nor the version number observed.

File: src/Framework/test/TargetingPackTests.cs [hunk 3]
```diff
@@ -177,5 +201,60 @@ public void PlatformManifestListsAllFiles()
...
+            Assert.Empty(missing);
+            Assert.Empty(unexpected);
+
+            Assert.All(frameworkListEntries, i =>
+            {
+                var assemblyVersion = i.Attribute("AssemblyVersion").Value;
+                var fileVersion = i.Attribute("FileVersion").Value;
+
+                Assert.True(Version.TryParse(assemblyVersion, out _), "Assembly version must be convertable to System.Version");
+                Assert.True(Version.TryParse(fileVersion, out _), "File version must be convertable to System.Version");
```
Comment (dougbu) [inline src/Framework/test/TargetingPackTests.cs:256]: Suggest adding more detail to these messages too

**Review by dougbu — APPROVED** [review]
Minor nits. Separately, the new tests check against _calculated_ expectations. Do we check the expected assemblies against a hard-coded list anywhere to make sure e.g. `_TargetingPackDependencies` was calculated correctly?

**Comment by JunTaoLuo** [comment]
I checked our tests in the shared framework and the targeting pack and it seems we never use any hardcoded values for expected assemblies. Is that something we want to add?

File: src/Framework/test/TargetingPackTests.cs (hunk 2 above)
Comment (JunTaoLuo) [reply src/Framework/test/TargetingPackTests.cs:56]: It does, path would include the name of the assembly. I could add the version to the message though.

**Comment by Pilchie** [comment]
I think it would be good to have a hardcoded baseline that we test against so that we notice when something is added/removed from the targeting pack.

**Pull request merged**